modules/dcache/src/main/java/org/dcache/pool/statistics/IoStatisticsChannel.java

```diff
@@ -0,0 +1,168 @@
+import java.io.SyncFailedException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * This class decorates any RepositoryChannel and adds the function of collecting data for the IoStatistics
+ */
+public class IoStatisticsChannel implements RepositoryChannel {
+
+    private static final Logger LOGGER =
+            LoggerFactory.getLogger(IoStatisticsChannel.class);
+
+    /**
+     * Inner channel to which all operations are delegated.
+     */
+    @VisibleForTesting
+    private RepositoryChannel _channel;
+
+    private final IoStatistics _statistics = new IoStatistics();
+
+    public IoStatisticsChannel(RepositoryChannel channel){ _channel = channel; }
+
+    /**
+     * Returns the object most central of this decorator
+     * @return object with collected and evaluated statistics data
+     */
+    public IoStatistics getStatistics() {
+        return _statistics;
+    }
+
+    @Override
+    public int write(ByteBuffer buffer, long position) throws IOException {
+        long startTime = System.currentTimeMillis();
+        int writtenBytes = _channel.write(buffer, position); // might be 0 if nothing has been written
+        long duration = System.currentTimeMillis() - startTime;
+        _statistics.updateWrite(writtenBytes, duration);
+        return writtenBytes;
+    }
+
+    @Override
+    public int read(ByteBuffer buffer, long position) throws IOException {
+        long startTime = System.currentTimeMillis();
+        int readBytes = _channel.read(buffer, position); // -1 => position greater than file; 0 => if end of file
+        long duration = System.currentTimeMillis() - startTime;
+        _statistics.updateRead(readBytes, duration);
+        return readBytes;
+    }
+
+    @Override
+    public void sync() throws SyncFailedException, IOException {
+        _channel.sync();
+    }
+
+    @Override
+    public long transferTo(long position, long count, WritableByteChannel target) throws IOException {
+        long startTime = System.currentTimeMillis();
+        long readBytes =  _channel.transferTo(position, count, target);
+        long duration = System.currentTimeMillis() - startTime;
+        _statistics.updateRead(readBytes, duration);
+        // long discrepance = count - readBytes;
+        return readBytes;
+    }
+
+    @Override
+    public long transferFrom(ReadableByteChannel src, long position, long count) throws IOException {
+        long startTime = System.currentTimeMillis();
+        long writtenBytes =  _channel.transferFrom(src, position, count);
+        long duration = System.currentTimeMillis() - startTime;
+        _statistics.updateWrite(writtenBytes, duration);
+        // long discrepance = count - writtenBytes;
+        return writtenBytes;
+    }
+
+    @Override
+    public long write(ByteBuffer[] srcs, int offset, int length) throws IOException {
```

Comment:
It seems you are not actually implementing the delegated method here.

long writtenBytes = _channel.write(srcs, offset, length) ?